File: frontend/lib/services/patient_service.dart
import 'package:dio/dio.dart';
import 'package:image_picker/image_picker.dart';
import '../core/constants/api_constants.dart';
import '../models/patient.dart';
import 'auth_service.dart';
import 'dio_client.dart';
import 'network_checker.dart';

class PatientService {
  final AuthService authService = AuthService();

  // -------------------------
  // Get Patients
  // -------------------------
  Future<Map<String, dynamic>> getPatients({
    String? query,
    int page = 1,
    int limit = 1000,
  }) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();
      if (token == null) {
        return {'success': false, 'message': 'غير مسجل الدخول'};
      }

      String url =
          '${ApiConstants.baseUrl}${ApiConstants.patients}?page=$page&limit=$limit';

      if (query != null && query.isNotEmpty) {
        url += '&q=${Uri.encodeQueryComponent(query)}';
      }

      final response = await DioClient.dio.get(
        url,
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      final data = response.data;
      final patients = (data['data'] as List)
          .map((e) => Patient.fromJson(e))
          .toList();

      return {
        'success': true,
        'patients': patients,
        'pagination': data['pagination'],
      };
    } catch (e) {
      return {
        'success': false,
        'message': 'تعذر الاتصال بالخادم',
      };
    }
  }

  // -------------------------
  // Get All Patients
  // -------------------------
  Future<Map<String, dynamic>> getAllPatients({String? query}) async {
    final List<Patient> all = [];
    int page = 1;
    int size = 100;

    while (true) {
      final result =
          await getPatients(query: query, page: page, limit: size);

      if (result['success'] != true) {
        return result;
      }

      final batch = result['patients'] as List<Patient>;
      all.addAll(batch);

      final pagination = result['pagination'];
      bool hasMore = false;

      if (pagination is Map && pagination['has_next'] == true) {
        hasMore = true;
      } else if (batch.length == size) {
        hasMore = true;
      }

      if (!hasMore) break;
      page++;
    }

    return {'success': true, 'patients': all};
  }

  // -------------------------
  // Get Patient By ID
  // -------------------------
  Future<Patient?> getPatient(String id) async {
    if (!await NetworkChecker.hasInternet()) return null;

    try {
      final token = await authService.getToken();
      if (token == null) return null;

      final response = await DioClient.dio.get(
        '${ApiConstants.baseUrl}${ApiConstants.patientById(id)}',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      return Patient.fromJson(response.data['data']);
    } catch (_) {
      return null;
    }
  }

  // -------------------------
  // Create Patient
  // -------------------------
  Future<Map<String, dynamic>> createPatient({
    required String name,
    required String phone,
    required String address,
  }) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();
      if (token == null) {
        return {'success': false, 'message': 'غير مسجل الدخول'};
      }

      final response = await DioClient.dio.post(
        '${ApiConstants.baseUrl}${ApiConstants.patients}',
        data: {
          'name': name,
          'phone': phone,
          'address': address,
        },
        options: Options(
          headers: {'Authorization': 'Bearer $token'},
        ),
      );

      return {
        'success': true,
        'patient': Patient.fromJson(response.data['data']),
      };
    } catch (e) {
      return {'success': false, 'message': 'فشل الاتصال بالخادم'};
    }
  }

  // -------------------------
  // Update Patient
  // -------------------------
  Future<Map<String, dynamic>> updatePatient({
    required String id,
    required String name,
    required String phone,
    required String address,
    String? note,
  }) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();

      final response = await DioClient.dio.patch(
        '${ApiConstants.baseUrl}${ApiConstants.patientById(id)}',
        data: {
          'name': name,
          'phone': phone,
          'address': address,
          if (note != null) 'note': note,
        },
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      return {
        'success': true,
        'patient': Patient.fromJson(response.data['data']),
      };
    } catch (_) {
      return {'success': false, 'message': 'فشل تحديث بيانات المريض'};
    }
  }

  // -------------------------
  // Upload Images
  // -------------------------
  Future<Map<String, dynamic>> uploadImages({
    required String patientId,
    required int stepNumber,
    required List<XFile> images,
  }) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();
      final url =
          '${ApiConstants.baseUrl}${ApiConstants.uploadImages(patientId, stepNumber)}';

      final formData = FormData();

      for (var img in images) {
        formData.files.add(
          MapEntry(
            'files',
            await MultipartFile.fromFile(img.path),
          ),
        );
      }

      final response = await DioClient.dio.post(
        url,
        data: formData,
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      return {
        'success': true,
        'message': response.data['message'] ?? 'تم رفع الصور بنجاح',
        'data': response.data['data'],
      };
    } catch (_) {
      return {
        'success': false,
        'message': 'فشل رفع الصور، حاول مرة أخرى',
      };
    }
  }

  // -------------------------
  // Mark Step Done
  // -------------------------
  Future<Map<String, dynamic>> markStepDone({
    required String patientId,
    required int stepNumber,
    required bool isDone,
  }) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();

      await DioClient.dio.patch(
        '${ApiConstants.baseUrl}${ApiConstants.markStepDone(
              patientId,
              stepNumber,
            )}',
        data: {'is_done': isDone},
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      return {'success': true};
    } catch (_) {
      return {'success': false, 'message': 'فشل تحديث الخطوة'};
    }
  }

  // -------------------------
  // Delete Patient
  // -------------------------
  Future<Map<String, dynamic>> deletePatient(String id) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();
      if (token == null) {
        return {'success': false, 'message': 'غير مسجل الدخول'};
      }

      await DioClient.dio.delete(
        '${ApiConstants.baseUrl}${ApiConstants.patientById(id)}',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      return {'success': true};
    } catch (_) {
      return {'success': false, 'message': 'فشل حذف المريض'};
    }
  }

  // -------------------------
  // Completed Patients (Admin)
  // -------------------------
  Future<Map<String, dynamic>> getCompletedPatients() async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();
      if (token == null) {
        return {'success': false, 'message': 'غير مسجل الدخول'};
      }

      final response = await DioClient.dio.get(
        '${ApiConstants.baseUrl}${ApiConstants.patients}/filter/completed',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      final patients = (response.data['data'] as List)
          .map((e) => Patient.fromJson(e))
          .toList();

      return {'success': true, 'patients': patients};
    } catch (_) {
      return {'success': false, 'message': 'فشل جلب المرضى المكتملين'};
    }
  }

  // -------------------------
  // Delete Image
  // -------------------------
  Future<Map<String, dynamic>> deleteImage({
    required String patientId,
    required int stepNumber,
    required String imageId,
  }) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();

      await DioClient.dio.delete(
        '${ApiConstants.baseUrl}${ApiConstants.deleteImage(
              patientId,
              stepNumber,
              imageId,
            )}',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      return {'success': true, 'message': 'تم حذف الصورة'};
    } catch (_) {
      return {'success': false, 'message': 'فشل حذف الصورة'};
    }
  }

  // -------------------------
  // Dashboard Stats
  // -------------------------
  Future<Map<String, dynamic>> getStatistics() async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();

      final response = await DioClient.dio.get(
        '${ApiConstants.baseUrl}/patients/stats/dashboard',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      return {
        'success': true,
        'data': response.data['data'],
      };
    } catch (_) {
      return {'success': false, 'message': 'فشل جلب الإحصائيات'};
    }
  }

  // -------------------------
  // Completed by Phase
  // -------------------------
  Future<Map<String, dynamic>> getCompletedByPhase(int phase) async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();
      if (token == null) {
        return {'success': false, 'message': 'غير مسجل الدخول'};
      }

      final response = await DioClient.dio.get(
        '${ApiConstants.baseUrl}/patients/filter/completed/phase/$phase',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      final patients = (response.data['data'] as List)
          .map((e) => Patient.fromJson(e))
          .toList();

      return {'success': true, 'patients': patients};
    } catch (_) {
      return {'success': false, 'message': 'فشل جلب المرضى للمراحل'};
    }
  }

  // -------------------------
  // Zero-step patients
  // -------------------------
  Future<Map<String, dynamic>> getZeroStepPatients() async {
    if (!await NetworkChecker.hasInternet()) {
      return {'success': false, 'message': 'لا يوجد اتصال بالإنترنت'};
    }

    try {
      final token = await authService.getToken();
      if (token == null) {
        return {'success': false, 'message': 'غير مسجل الدخول'};
      }

      final response = await DioClient.dio.get(
        '${ApiConstants.baseUrl}/patients/filter/zero-step',
        options: Options(headers: {'Authorization': 'Bearer $token'}),
      );

      final patients = (response.data['data'] as List)
          .map((e) => Patient.fromJson(e))
          .toList();

      return {'success': true, 'patients': patients};
    } catch (_) {
      return {'success': false, 'message': 'فشل جلب المرضى (لا خطوات مكتملة)'};
    }
  }
}

File: backend/app/services/patient_service.py
from typing import List, Optional, Tuple
from fastapi import HTTPException

from ..models.patient import Patient, Step, Image
from ..schemas.patient_schema import PatientCreate


STEP_TITLES: List[str] = [
    "صورة وجه كامل للمريض",
    "صورة ابتسامة المريض بدون وجه",
    "صورة ابتسامة مع ريتراكتد",
    "صورة ابتسامة مع ريتراكتد يمين",
    "صورة ابتسامة مع ريتراكتد يسار",
    "صورة اللهاة العلوية",
    "صورة اللهاة السفلية",
    "صورة الأشعة قبل الزراعة",
    "مكان الزراعة قبل الشق",
    "مكان الزراعة بعد الشق",
    "وضع الدرل",
    "الزرعة داخل الفم",
    "صورة الخيط",
    "الأشعة بعد الزراعة",
    "صورة الطبعة",
    "التركيب على القالب",
    "التركيب داخل الفم ابتسامة",
    "التركيب داخل الفم يمين",
    "التركيب داخل الفم يسار",
    "اللهاة العلوية",
    "اللهاة السفلية",
    "صورة وجه كامل",
    "صورة اشعة سنسر pa و اشعة التركيب",  # step 23 (after operation)
    "صورة المعالجة",  # step 24 (treatment)
    "صورة Emergency profile",  # step 25 (after operation - new)
]


def generate_default_steps() -> List[Step]:
    steps: List[Step] = []
    for i, title in enumerate(STEP_TITLES, start=1):
        steps.append(Step(step_number=i, title=title))
    return steps


async def create_patient(data: PatientCreate) -> Patient:
    patient = Patient(
        name=data.name,
        phone=data.phone,
        address=data.address,
        note=data.note,
        steps=generate_default_steps(),
    )
    await patient.insert()
    return patient


async def list_patients() -> List[Patient]:
    return await Patient.find_all().to_list()


def _phase_steps(phase: int) -> List[int]:
    if phase == 1:
        return [1, 2, 3, 4, 5, 6, 7, 8]
    if phase == 2:
        return [9, 10, 11, 12, 13, 14]
    if phase == 3:
        # treatment
        return [24]
    if phase == 4:
        # after operation (includes newly added 25)
        return [15, 16, 17, 18, 19, 20, 21, 22, 23, 25]
    raise HTTPException(status_code=400, detail="Invalid phase")


def _is_phase_completed(patient: Patient, phase: int) -> bool:
    required = set(_phase_steps(phase))
    step_map = {s.step_number: s for s in patient.steps}
    # All required steps must exist and be done
    for sn in required:
        step = step_map.get(sn)
        if step is None or not step.is_done:
            return False
    return True


def _has_zero_steps_done(patient: Patient) -> bool:
    return not any(s.is_done for s in patient.steps)


async def search_patients(q: Optional[str], page: int, limit: int) -> dict:
    skip = (page - 1) * limit
    if q:
        filter_query = {
            "$or": [
                {"name": {"$regex": q, "$options": "i"}},
                {"phone": {"$regex": q, "$options": "i"}},
            ]
        }
        cursor = Patient.find(filter_query)
        total = await Patient.find(filter_query).count()
    else:
        cursor = Patient.find_all()
        total = await Patient.find_all().count()

    patients = await cursor.skip(skip).limit(limit).to_list()

    pages = (total + limit - 1) // limit if total else 0
    return {
        "data": patients,
        "pagination": {
            "total": total,
            "page": page,
            "limit": limit,
            "pages": pages,
        },
        "query": q,
    }


async def get_patient(patient_id: str) -> Patient:
    patient = await Patient.get(patient_id)
    if not patient:
        raise HTTPException(status_code=404, detail="Patient not found")
    return patient


async def update_patient(patient_id: str, name: str, phone: str, address: str, note: Optional[str] = None) -> Patient:
    """تحديث بيانات المريض الأساسية"""
    patient = await get_patient(patient_id)
    patient.name = name
    patient.phone = phone
    patient.address = address
    if note is not None:
        patient.note = note
    await patient.save()
    return patient


def get_step(patient: Patient, step_number: int) -> Step:
    for step in patient.steps:
        if step.step_number == step_number:
            return step
    raise HTTPException(status_code=404, detail="Step not found")


async def mark_step_done(patient_id: str, step_number: int, is_done: bool = True) -> Patient:
    patient = await get_patient(patient_id)
    step = get_step(patient, step_number)
    step.is_done = is_done
    await patient.save()
    return patient


async def add_images_to_step(patient_id: str, step_number: int, images: List[Image]) -> Patient:
    patient = await get_patient(patient_id)
    step = get_step(patient, step_number)
    step.images.extend(images)
    await patient.save()
    return patient


async def delete_image_from_step(patient_id: str, step_number: int, image_id: str) -> Patient:
    patient = await get_patient(patient_id)
    step = get_step(patient, step_number)
    before = len(step.images)
    step.images = [img for img in step.images if img.id != image_id]
    if len(step.images) == before:
        raise HTTPException(status_code=404, detail="Image not found")
    await patient.save()
    return patient


async def get_statistics() -> dict:
    """جلب إحصائيات المرضى"""
    all_patients = await Patient.find_all().to_list()
    total = len(all_patients)

    completed_total = 0
    incomplete_total = 0
    zero_step_total = 0

    # Phase completion counters
    phase_completed = {1: 0, 2: 0, 3: 0, 4: 0}

    for patient in all_patients:
        all_steps_done = all(step.is_done for step in patient.steps)
        if all_steps_done:
            completed_total += 1
        else:
            incomplete_total += 1

        if _has_zero_steps_done(patient):
            zero_step_total += 1

        for phase in (1, 2, 3, 4):
            if _is_phase_completed(patient, phase):
                phase_completed[phase] += 1

    return {
        "total_patients": total,
        "completed_patients": completed_total,
        "incomplete_patients": incomplete_total,
        "zero_step_patients": zero_step_total,
        "phase_completed": phase_completed,
    }


async def get_completed_patients() -> List[Patient]:
    """جلب فقط المرضى المكتملين"""
    all_patients = await Patient.find_all().to_list()
    completed_patients = []
    
    for patient in all_patients:
        if all(step.is_done for step in patient.steps):
            completed_patients.append(patient)
    
    return completed_patients


async def delete_patient(patient_id: str) -> None:
    """حذف مريض"""
    patient = await get_patient(patient_id)
    await patient.delete()


# ========== Phase/Zero-step filters ==========

async def get_patients_completed_phase(phase: int) -> List[Patient]:
    """جلب المرضى الذين أتمّوا المرحلة المحددة (كل خطوات المرحلة مكتملة)"""
    all_patients = await Patient.find_all().to_list()
    result: List[Patient] = []
    for patient in all_patients:
        if _is_phase_completed(patient, phase):
            result.append(patient)
    return result


async def get_zero_step_patients() -> List[Patient]:
    """المرضى الذين لم يكملوا أي خطوة"""
    all_patients = await Patient.find_all().to_list()
    return [p for p in all_patients if _has_zero_steps_done(p)]
